bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttWillAndTestament.java

```diff
@@ -45,39 +45,39 @@ public static MqttWillAndTestament fromString(String string) {
             String value = StringUtils.trimToEmpty(components[i]);
             switch (i) {
                 case 0:
-                    result.setTopic(value);
+                    result.topic = value;
                     break;
                 case 1:
-                    result.setPayload(value.getBytes());
+                    result.payload = value.getBytes();
                     break;
                 case 2:
                     if (!"".equals(value)) {
-                        result.setQos(Integer.valueOf(value));
+                        int qos = Integer.valueOf(value);
+                        if (qos >= 0 && qos <= 2) {
+                            result.qos = qos;
+                        }
                     }
                     break;
                 case 3:
-                    result.setRetain(Boolean.valueOf(value));
+                    result.retain = Boolean.valueOf(value);
                     break;
             }
         }
-        return result;
-
+        return result.isValid() ? result : null;
     }
 
     /**
-     * @return the topic for the last will.
+     * Return true if the last will and testament object is valid.
      */
-    public String getTopic() {
-        return topic;
+    private boolean isValid() {
+        return !StringUtils.isBlank(topic) && payload.length > 0;
```

Comment:
It's been some time since I read the spec and wrote a MQTT broker, so excuse me if it is not correct, but if IIRC:
If the retained flag is used for MQTT messages, a null message is used to "un"-retain a topic again.
So using this approach to signal a disconnect (remove a retained topic) is a valid approach.

Miss I something?